src/core/credentials/transport/tls/spiffe_utils.h

```diff
@@ -45,6 +47,109 @@ class SpiffeId final {
   const std::string path_;
 };
 
+// An entry in the Key vector of a SPIFFE Bundle following these documents:
+// https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#3-spiffe-bundles
+// https://github.com/grpc/proposal/blob/master/A87-mtls-spiffe-support.md
+class SpiffeBundleKey final {
+ public:
+  static const JsonLoaderInterface* JsonLoader(const JsonArgs&) {
+    static const auto* kLoader =
+        JsonObjectLoader<SpiffeBundleKey>()
+            .Field("kty", &SpiffeBundleKey::kty_)
+            .OptionalField("kid", &SpiffeBundleKey::kid_)
+            .Field("use", &SpiffeBundleKey::use_)
+            .Field("x5c", &SpiffeBundleKey::x5c_)
+            .Field("n", &SpiffeBundleKey::n_)
+            .Field("e", &SpiffeBundleKey::e_)
+            .Finish();
+    return kLoader;
+  }
+
+  void JsonPostLoad(const Json& json, const JsonArgs&,
+                    ValidationErrors* errors);
+
+  // Returns the PEM x509 string for this SPIFFE Bundle entry which is the only
+  // entry in the x5c_ vector.
+  absl::StatusOr<absl::string_view> GetRoot();
+
+ private:
+  // The following are all fields of the the JWK key in the SPIFFE bundle per
+  // https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#3-spiffe-bundles
+  std::string kty_;
+  std::string kid_;
+  std::string use_;
+  std::vector<std::string> x5c_;
+  std::string n_;
+  std::string e_;
+};
+
+// A SPIFFE bundle consists of a trust domain and a set of roots for that trust
+// domain.
+// https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md#3-spiffe-bundles
+// https://github.com/grpc/proposal/blob/master/A87-mtls-spiffe-support.md
+class SpiffeBundle final {
+ public:
+  static const JsonLoaderInterface* JsonLoader(const JsonArgs&) {
+    static const auto* kLoader = JsonObjectLoader<SpiffeBundle>()
+                                     .Field("keys", &SpiffeBundle::keys_)
+                                     .Finish();
+    return kLoader;
+  }
+
+  // Returns a vector of the roots in this SPIFFE Bundle.
+  absl::StatusOr<std::vector<absl::string_view>> GetRoots();
+
+ private:
+  std::vector<SpiffeBundleKey> keys_;
+};
+
+// A map of SPIFFE bundles keyed to trust domains. This functions as a map of a
+// given trust domain to the root certificates that should be used when
+// validating certificates in this trust domain.
+// https://github.com/grpc/proposal/blob/master/A87-mtls-spiffe-support.md
+// https://github.com/grpc/proposal/blob/master/A87-mtls-spiffe-support.md
+// Only configuring X509 roots is supported.
+class SpiffeBundleMap final {
+ public:
+  static const JsonLoaderInterface* JsonLoader(const JsonArgs&) {
+    static const auto* kLoader =
+        JsonObjectLoader<SpiffeBundleMap>()
+            .Field("trust_domains", &SpiffeBundleMap::temp_bundles_)
+            .Finish();
+    return kLoader;
+  }
+
+  void JsonPostLoad(const Json& json, const JsonArgs&,
```

Comment:
Should this be private?

Reply:
No, I don't believe any of the json methods can be private as they are called by the json libraries